parlai/core/torch_ranker_agent.py

```diff
@@ -308,30 +319,83 @@ def _get_model_file(self, opt):
-                print("Vectorizing fixed candidates set from {}".format(
-                    self.opt['fixed_candidates_path']))
-                with open(self.opt['fixed_candidates_path'], 'r') as f:
-                    self.fixed_candidates = [line.strip() for line in
-                        f.readlines()]
-                vectorize_func = self._vectorize_text
-                candidate_vecs = [vectorize_func(cand) for cand in
-                    self.fixed_candidates]
-                self.fixed_candidates_vec, _ = padded_tensor(candidate_vecs,
-                    use_cuda=self.use_cuda)
+            opt = self.opt
+            if opt['fixed_candidates_path']:
+                print("[ Loading fixed candidate set text from {} ]".format(
+                    opt['fixed_candidates_path']))
+                with open(opt['fixed_candidates_path'], 'r') as f:
+                    self.fixed_candidates = [line.strip() for line in f.readlines()]
+
+                if (opt['fixed_candidate_vecs_path'] and
+                        os.path.isfile(opt['fixed_candidate_vecs_path'])):
+                    print("[ Loading fixed candidate set vectors from {} ]".format(
+                        opt['fixed_candidate_vecs_path']))
+                    self.fixed_candidate_vecs = torch.load(
+                        opt['fixed_candidate_vecs_path'])
+                else:
+                    cands = self.fixed_candidates
+                    cand_batches = [cands[i:i + 512] for i in range(0, len(cands), 512)]
+                    print("[ Vectorizing fixed candidates set from {} ({} batch(es) of "
+                          "up to 512) ]".format(opt['fixed_candidates_path'],
+                                          len(cand_batches)))
+                    cand_vecs = []
+                    for batch in cand_batches:
+                        cand_vecs.extend(self.vectorize_fixed_candidates(batch))
+                    self.fixed_candidate_vecs = self._cat_and_pad(cand_vecs,
+                                                                  self.NULL_IDX)
+                    if opt['fixed_candidate_vecs_path']:
+                        print("[ Saving fixed candidate set vectors to {} ]".format(
+                            opt['fixed_candidate_vecs_path']))
+                        torch.save(self.fixed_candidate_vecs,
+                                   opt['fixed_candidate_vecs_path'])
+
+                if self.use_cuda:
+                    self.fixed_candidate_vecs = self.fixed_candidate_vecs.cuda()
             else:
                 self.fixed_candidates = None
-                self.fixed_candidates_vec = None
+                self.fixed_candidate_vecs = None
+
+    def vectorize_fixed_candidates(self, cands_batch):
+        return [self._vectorize_text(cand, truncate=self.truncate, truncate_left=False)
+                for cand in cands_batch]
+
+    def _cat_and_pad(self, tensor_list, max_len=None, use_cuda=False):
```

Comment:
use `parlai.core.utils.padded_tensor` instead

Reply:
the types and dimensions don't line up: it expects a list of lists, and we have a list of list of tensors.

Reply:
that's what parlai.core.utils.padded_3d is for